pubsub/streaming-analytics/PubSubToGCS.py

```diff
@@ -14,119 +14,126 @@
 
 # [START pubsub_to_gcs]
 import argparse
-import datetime
-import json
+from datetime import datetime
 import logging
+import random
 
-import apache_beam as beam
+from apache_beam import DoFn, GroupByKey, io, ParDo, Pipeline, PTransform, WindowInto
 from apache_beam.options.pipeline_options import PipelineOptions
-import apache_beam.transforms.window as window
+from apache_beam.transforms.window import FixedWindows
 
 
-class GroupWindowsIntoBatches(beam.PTransform):
-    """A composite transform that groups Pub/Sub messages based on publish
-    time and outputs a list of dictionaries, where each contains one message
-    and its publish timestamp.
+class GroupMessagesByFixedWindows(PTransform):
+    """A composite transform that groups Pub/Sub messages based on publish time
+    and outputs a list of tuples, each containing a message and its publish time.
     """
 
-    def __init__(self, window_size):
-        # Convert minutes into seconds.
+    def __init__(self, window_size, num_shards):
```

Comment:
Since `num_shards` is more of an optimization parameter, how about we have a default value even if we don't use it? So if anyone wants to copy-paste this class, it would be a little easier to use. Elasticsearch uses 5 as a default, maybe we can use that as well or whatever you prefer. I would just avoid having 1 as a default so that we "optimize by default" and to showcase this as a feature.

Reply:
Good point. Setting default to 5.